File: jscomp/others/belt_MutableMap.ml

(* Copyright (C) 2017 Authors of BuckleScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module Int = Belt_MutableMapInt
module String = Belt_MutableMapString
    
module N = Belt_internalAVLtree
module A = Belt_Array


type ('key, 'id) id = ('key, 'id) Belt_Id.comparable
type ('key, 'id ) cmp = ('key, 'id) Belt_Id.cmp
    
module S = struct    
  type ('k, 'v, 'id) t = {
    cmp: ('k, 'id) cmp;
    mutable data: ('k, 'v) N.t
  } [@@bs.deriving abstract]
end

type ('k, 'v, 'id) t = ('k, 'v, 'id) S.t

let rec removeMutateAux nt x ~cmp = 
  let k = N.keyGet nt in 
  let c = (Belt_Id.getCmpInternal cmp) x k [@bs] in 
  if c = 0 then 
    let l,r = N.(leftGet nt, rightGet nt) in       
    match N.(toOpt l, toOpt r) with 
    | Some _,  Some nr ->  
      N.rightSet nt (N.removeMinAuxWithRootMutate nt nr);
      N.return (N.balMutate nt)
    | None, Some _ ->
      r  
    | (Some _ | None ), None ->  l 
  else 
    begin 
      if c < 0 then 
        match N.toOpt (N.leftGet nt) with         
        | None -> N.return nt 
        | Some l ->
          N.leftSet nt (removeMutateAux ~cmp l x );
          N.return (N.balMutate nt)
      else 
        match N.toOpt (N.rightGet nt) with 
        | None -> N.return nt 
        | Some r -> 
          N.rightSet nt (removeMutateAux ~cmp r x);
          N.return (N.balMutate nt)
    end    

let remove d k =  
  let oldRoot = S.dataGet d in   
  match N.toOpt oldRoot with 
  | None -> ()
  | Some oldRoot2 ->
    let newRoot = removeMutateAux ~cmp:(S.cmpGet d) oldRoot2 k in 
    if newRoot != oldRoot then 
      S.dataSet d newRoot    


let rec removeArrayMutateAux t xs i len ~cmp  =  
  if i < len then 
    let ele = A.getUnsafe xs i in 
    let u = removeMutateAux t ele ~cmp in 
    match N.toOpt u with 
    | None -> N.empty
    | Some t -> removeArrayMutateAux t xs (i+1) len ~cmp 
  else N.return t    

let removeMany d xs =  
  let oldRoot = S.dataGet d in 
  match N.toOpt oldRoot with 
  | None -> ()
  | Some nt -> 
    let len = A.length xs in 
    let newRoot = removeArrayMutateAux nt xs 0 len ~cmp:(S.cmpGet d) in 
    if newRoot != oldRoot then 
      S.dataSet d newRoot


let rec updateDone t x   f  ~cmp =   
  match N.toOpt t with 
  | None ->
    (match f None [@bs] with
    | Some data -> N.singleton x data
    | None -> t)
  | Some nt -> 
    let k = N.keyGet nt in 
    let  c = (Belt_Id.getCmpInternal cmp) x k [@bs] in  
    if c = 0 then begin     
      match f (Some (N.valueGet nt)) [@bs] with
      | None ->
        let l,r = N.leftGet nt, N.rightGet nt in
        begin match N.toOpt l, N.toOpt r with
        | Some _, Some nr ->
          N.rightSet nt (N.removeMinAuxWithRootMutate nt nr);
          N.return (N.balMutate nt)
        | None, Some _ ->
          r
        | (Some _ | None), None -> l
        end
      | Some data -> 
        N.valueSet nt data;
        N.return nt
    end      
    else
      let l, r = N.(leftGet nt, rightGet nt) in 
      (if c < 0 then                   
         let ll = updateDone  l x f ~cmp in
         N.leftSet nt ll
       else   
         N.rightSet nt (updateDone  r x f ~cmp);
      );
      N.return (N.balMutate nt)  
      
let updateU t  x f =       
  let oldRoot = S.dataGet t in 
  let newRoot = updateDone oldRoot x f ~cmp:(S.cmpGet t) in 
  if  newRoot != oldRoot then 
    S.dataSet t newRoot 
let update t x f = updateU t x (fun [@bs] a -> f a)
    
let make (type key) (type identity) ~(id : (key,identity) id) =
  let module M = (val id) in 
  S.t ~cmp:M.cmp ~data:N.empty

let clear m = S.dataSet m N.empty
    
let isEmpty d = 
  N.isEmpty (S.dataGet d)
    

let minKey m = N.minKey (S.dataGet m)
let minKeyUndefined m = N.minKeyUndefined (S.dataGet m)
let maxKey m = N.maxKey (S.dataGet m)
let maxKeyUndefined m = N.maxKeyUndefined (S.dataGet m)
let minimum m = N.minimum (S.dataGet m)
let minUndefined m = N.minUndefined (S.dataGet m) 
let maximum m = N.maximum (S.dataGet m)
let maxUndefined m = N.maxUndefined (S.dataGet m)

let forEachU d f = N.forEachU (S.dataGet d) f
let forEach d f = forEachU d (fun [@bs] a b -> f a b)
let reduceU d acc cb = N.reduceU (S.dataGet d) acc cb
let reduce d acc cb = reduceU d acc (fun[@bs] a b c -> cb a b c)
let everyU d p = N.everyU (S.dataGet d) p
let every d p = everyU d (fun[@bs] a b -> p a b)
let someU d p = N.someU (S.dataGet d) p       
let some d p = someU d (fun [@bs] a b -> p a b)
let size d = 
  N.size (S.dataGet d)
let toList d =
  N.toList (S.dataGet d)
let toArray d = 
  N.toArray (S.dataGet d)
let keysToArray d =   
  N.keysToArray (S.dataGet d)
let valuesToArray d =   
  N.valuesToArray (S.dataGet d)
    
let fromSortedArrayUnsafe (type key) (type identity) ~(id : (key,identity) id) xs : _ t =
  let module M = (val id) in 
  S.t ~data:(N.fromSortedArrayUnsafe xs) ~cmp:M.cmp
    
let checkInvariantInternal d = 
  N.checkInvariantInternal (S.dataGet d)  

let cmpU m1 m2 cmp = 
  N.cmpU ~kcmp:(S.cmpGet m1) ~vcmp:cmp (S.dataGet m1) (S.dataGet m2)
let cmp m1 m2 cmp = cmpU m1 m2 (fun[@bs] a b -> cmp a b)
    
let eqU m1 m2 cmp = 
  N.eqU ~kcmp:(S.cmpGet m1) ~veq:cmp (S.dataGet m1) (S.dataGet m2)
let eq m1 m2 cmp = eqU m1 m2 (fun[@bs] a b -> cmp a b)

let mapU m f = 
  S.t ~cmp:(S.cmpGet m) ~data:(N.mapU (S.dataGet m) f)
let map m f = mapU m (fun[@bs] a  -> f a )    
let mapWithKeyU m f = 
  S.t ~cmp:(S.cmpGet m) ~data:(N.mapWithKeyU (S.dataGet m) f)
let mapWithKey m f = mapWithKeyU m (fun [@bs] a b -> f a b)     
let get m x  = 
  N.get ~cmp:(S.cmpGet m)  (S.dataGet m) x
    
let getUndefined m x = 
  N.getUndefined ~cmp:(S.cmpGet m)  (S.dataGet m) x
    
let getWithDefault m x def = 
  N.getWithDefault ~cmp:(S.cmpGet m) (S.dataGet m) x  def
    
let getExn m x = 
  N.getExn ~cmp:(S.cmpGet m) (S.dataGet m) x
    
let has m x = 
  N.has ~cmp:(S.cmpGet m) (S.dataGet m) x
    
let fromArray (type k) (type identity) data ~(id : (k,identity) id)= 
  let module M = (val id ) in
  let cmp = M.cmp in 
  S.t ~cmp  ~data:(N.fromArray ~cmp data)
  
let set  m e v = 
  let oldRoot = S.dataGet m in 
  let newRoot = N.updateMutate ~cmp:(S.cmpGet m) oldRoot e v in 
  if newRoot != oldRoot then 
    S.dataSet m newRoot

let mergeManyAux t  xs ~cmp =     
  let v = ref t in 
  for i = 0 to A.length xs - 1 do 
    let key,value = A.getUnsafe xs i in 
    v .contents<- N.updateMutate v.contents key value ~cmp
  done; 
  v.contents 

let mergeMany d xs =   
  let oldRoot = S.dataGet d in 
  let newRoot = mergeManyAux oldRoot xs ~cmp:(S.cmpGet d) in 
  if newRoot != oldRoot then 
    S.dataSet d newRoot

File: lib/es6/Belt_MutableMap.js


import * as Primitive_option from "./Primitive_option.js";
import * as Belt_internalAVLtree from "./Belt_internalAVLtree.js";

function removeMutateAux(nt, x, cmp) {
  let k = nt.k;
  let c = cmp(x, k);
  if (c === 0) {
    let l = nt.l;
    let r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLtree.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLtree.balMutate(nt);
      } else {
        return l;
      }
    } else if (r !== undefined) {
      return r;
    } else {
      return l;
    }
  }
  if (c < 0) {
    let l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = removeMutateAux(l$1, x, cmp);
      return Belt_internalAVLtree.balMutate(nt);
    } else {
      return nt;
    }
  }
  let r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = removeMutateAux(r$1, x, cmp);
    return Belt_internalAVLtree.balMutate(nt);
  } else {
    return nt;
  }
}

function remove(d, k) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let newRoot = removeMutateAux(oldRoot, k, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function removeArrayMutateAux(_t, xs, _i, len, cmp) {
  while (true) {
    let i = _i;
    let t = _t;
    if (i >= len) {
      return t;
    }
    let ele = xs[i];
    let u = removeMutateAux(t, ele, cmp);
    if (u === undefined) {
      return;
    }
    _i = i + 1 | 0;
    _t = u;
    continue;
  };
}

function removeMany(d, xs) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let len = xs.length;
  let newRoot = removeArrayMutateAux(oldRoot, xs, 0, len, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function updateDone(t, x, f, cmp) {
  if (t !== undefined) {
    let k = t.k;
    let c = cmp(x, k);
    if (c === 0) {
      let data = f(Primitive_option.some(t.v));
      if (data !== undefined) {
        t.v = Primitive_option.valFromOption(data);
        return t;
      }
      let l = t.l;
      let r = t.r;
      if (l !== undefined) {
        if (r !== undefined) {
          t.r = Belt_internalAVLtree.removeMinAuxWithRootMutate(t, r);
          return Belt_internalAVLtree.balMutate(t);
        } else {
          return l;
        }
      } else if (r !== undefined) {
        return r;
      } else {
        return l;
      }
    }
    if (c < 0) {
      t.l = updateDone(t.l, x, f, cmp);
    } else {
      t.r = updateDone(t.r, x, f, cmp);
    }
    return Belt_internalAVLtree.balMutate(t);
  }
  let data$1 = f(undefined);
  if (data$1 !== undefined) {
    return Belt_internalAVLtree.singleton(x, Primitive_option.valFromOption(data$1));
  } else {
    return t;
  }
}

function update(t, x, f) {
  let oldRoot = t.data;
  let newRoot = updateDone(oldRoot, x, f, t.cmp);
  if (newRoot !== oldRoot) {
    t.data = newRoot;
    return;
  }
  
}

function make(id) {
  return {
    cmp: id.cmp,
    data: undefined
  };
}

function clear(m) {
  m.data = undefined;
}

function isEmpty(d) {
  let x = d.data;
  return x === undefined;
}

function minKey(m) {
  return Belt_internalAVLtree.minKey(m.data);
}

function minKeyUndefined(m) {
  return Belt_internalAVLtree.minKeyUndefined(m.data);
}

function maxKey(m) {
  return Belt_internalAVLtree.maxKey(m.data);
}

function maxKeyUndefined(m) {
  return Belt_internalAVLtree.maxKeyUndefined(m.data);
}

function minimum(m) {
  return Belt_internalAVLtree.minimum(m.data);
}

function minUndefined(m) {
  return Belt_internalAVLtree.minUndefined(m.data);
}

function maximum(m) {
  return Belt_internalAVLtree.maximum(m.data);
}

function maxUndefined(m) {
  return Belt_internalAVLtree.maxUndefined(m.data);
}

function forEach(d, f) {
  Belt_internalAVLtree.forEach(d.data, f);
}

function reduce(d, acc, cb) {
  return Belt_internalAVLtree.reduce(d.data, acc, cb);
}

function every(d, p) {
  return Belt_internalAVLtree.every(d.data, p);
}

function some(d, p) {
  return Belt_internalAVLtree.some(d.data, p);
}

function size(d) {
  return Belt_internalAVLtree.size(d.data);
}

function toList(d) {
  return Belt_internalAVLtree.toList(d.data);
}

function toArray(d) {
  return Belt_internalAVLtree.toArray(d.data);
}

function keysToArray(d) {
  return Belt_internalAVLtree.keysToArray(d.data);
}

function valuesToArray(d) {
  return Belt_internalAVLtree.valuesToArray(d.data);
}

function checkInvariantInternal(d) {
  Belt_internalAVLtree.checkInvariantInternal(d.data);
}

function cmp(m1, m2, cmp$1) {
  return Belt_internalAVLtree.cmp(m1.data, m2.data, m1.cmp, cmp$1);
}

function eq(m1, m2, cmp) {
  return Belt_internalAVLtree.eq(m1.data, m2.data, m1.cmp, cmp);
}

function map(m, f) {
  return {
    cmp: m.cmp,
    data: Belt_internalAVLtree.map(m.data, f)
  };
}

function mapWithKey(m, f) {
  return {
    cmp: m.cmp,
    data: Belt_internalAVLtree.mapWithKey(m.data, f)
  };
}

function get(m, x) {
  return Belt_internalAVLtree.get(m.data, x, m.cmp);
}

function getUndefined(m, x) {
  return Belt_internalAVLtree.getUndefined(m.data, x, m.cmp);
}

function getWithDefault(m, x, def) {
  return Belt_internalAVLtree.getWithDefault(m.data, x, def, m.cmp);
}

function getExn(m, x) {
  return Belt_internalAVLtree.getExn(m.data, x, m.cmp);
}

function has(m, x) {
  return Belt_internalAVLtree.has(m.data, x, m.cmp);
}

function fromArray(data, id) {
  let cmp = id.cmp;
  return {
    cmp: cmp,
    data: Belt_internalAVLtree.fromArray(data, cmp)
  };
}

function set(m, e, v) {
  let oldRoot = m.data;
  let newRoot = Belt_internalAVLtree.updateMutate(oldRoot, e, v, m.cmp);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
    return;
  }
  
}

function mergeManyAux(t, xs, cmp) {
  let v = t;
  for (let i = 0, i_finish = xs.length; i < i_finish; ++i) {
    let match = xs[i];
    v = Belt_internalAVLtree.updateMutate(v, match[0], match[1], cmp);
  }
  return v;
}

function mergeMany(d, xs) {
  let oldRoot = d.data;
  let newRoot = mergeManyAux(oldRoot, xs, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

let Int;

let $$String;

let cmpU = cmp;

let eqU = eq;

let forEachU = forEach;

let reduceU = reduce;

let everyU = every;

let someU = some;

let updateU = update;

let mapU = map;

let mapWithKeyU = mapWithKey;

export {
  Int,
  $$String,
  make,
  clear,
  isEmpty,
  has,
  cmpU,
  cmp,
  eqU,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  size,
  toList,
  toArray,
  fromArray,
  keysToArray,
  valuesToArray,
  minKey,
  minKeyUndefined,
  maxKey,
  maxKeyUndefined,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getWithDefault,
  getExn,
  checkInvariantInternal,
  remove,
  removeMany,
  set,
  updateU,
  update,
  mergeMany,
  mapU,
  map,
  mapWithKeyU,
  mapWithKey,
}
/* No side effect */
